network/local.go

```diff
@@ -306,36 +303,37 @@ func (c *connQueue) push(buff []byte) {
 // pop returns with an error if the queue is closed or gets closed while waiting
 // for a packet.
 func (c *connQueue) pop() ([]byte, error) {
-	c.L.Lock()
-	defer c.L.Unlock()
-	for len(c.queue) == 0 {
-		if c.closed {
+	select {
+	case buff, opened := <-c.queue:
+		if !opened {
 			return nil, ErrClosed
 		}
-		c.Wait()
-	}
-	if c.closed {
-		return nil, ErrClosed
+		return buff, nil
 	}
-	nm := c.queue[0]
-	c.queue = c.queue[1:]
-	return nm, nil
 }
 
 // close sets the closed-field to true and signals to all ongoing pop()
 // operations to return.
-func (c *connQueue) close() {
-	c.L.Lock()
-	defer c.L.Unlock()
-	c.closed = true
-	c.Broadcast()
+func (c *connQueue) close() error {
+	select {
+	case _, opened := <-c.queue:
+		if !opened {
+			return ErrClosed
+		}
+	default:
+		close(c.queue)
+	}
+	return nil
 }
 
 // isClosed returns whether this queue is closed or not.
 func (c *connQueue) isClosed() bool {
-	c.L.Lock()
-	defer c.L.Unlock()
-	return c.closed
+	select {
+	case _, opened := <-c.queue:
```

Comment:
What happens if a message is pending and `isClosed` is called? Won't the message be discarded?